parlai/core/teachers.py

```diff
@@ -1257,3 +1261,337 @@ def _setup_data(self, path):
+                % dt
+            )
+
+        # Assumes file is train.json or valid.json named
+        data_file = os.path.join(self.data_path, '%s.json' % dt)
+
+        # Load the text data and image number indexes
+        with open(data_file) as f:
+            self.data = json.load(f)
+
+        if len(self.data) > 0 and self.image_id_key not in self.data[0]:
+            # Data doesn't have a "image_id" like field so add the index in the file to the data
+            for idx, d in enumerate(self.data):
+                d[self.image_id_key] = idx
+
+        return self.data
+
+    def setup_data(self, data_path):
+        """Load text and image data.
+
+        The image features all live in dicts by default in <data_path>/image_features but get_image_features_path() above can be overriden by subclass to put them elsewhere.
+
+        In the (very odd) case that the resnet or resnext dicts (models buildable using ImageLoader) are not found, we build them.
+        """
+
+        image_model_features_dict_path = self.get_image_model_features_path(
+            self.task, self.image_model, self.datatype
+        )
+
+        if os.path.isfile(image_model_features_dict_path):
+            self.image_features_dict = torch.load(
+                image_model_features_dict_path, map_location='cpu'
+            )
+        else:
+            if self.is_image_model_buildable(self.image_model):
+                # try to build with ImageLoader (i.e. resenet/resnext variants)
+                self.image_features_dict = self._build_image_features_dict(
+                    self.data_path, self.datatype, image_model_features_dict_path
+                )
+            else:
+                raise RuntimeError(
+                    'Image model: %s is not buildable by ImageLoader but does not already exist on disk as an image features dict for this dataset.'
+                    % self.image_model
+                )
+
+    def _build_image_features_dict(self, data_path, dt, store_dict_path):
+        """ Build resne(x)t image features with ImageLoader
+        (Or anything handleable by ImageLoader) and save to path.
+        Only called if we haven't already built the dict before.
+
+        """
+        image_features_dict = {}
+        total = len(self.data)
+        import tqdm
+
+        pbar = tqdm.tqdm(
+            total=total,
+            unit='cand',
+            unit_scale=True,
+            desc='Building image features dict for %s with ImageLoader.'
+            % self.image_model,
+        )
+        num = 0
+        for ex in self.data:
+            img_id = ex[self.image_id_key]
+            img_path = self.image_id_to_image_path(img_id)
+            image = self.image_loader.load(img_path).detach()
+            image = image[0, :, 0, 0]
+            image_features_dict[img_id] = image
+            num += 1
+            pbar.update(1)
+            if num % 1000 == 0:
+                print('Processing image index: %s' % num)
```

Comment:
nit: Looks like most of this is coming from older code, but we could upgrade to f-strings. 

```python
print(f'Processing image index: {num}')
```

Reply:
Done.